filclient/filclient.go

```diff
@@ -853,221 +860,137 @@ func (fc *FilClient) RetrieveContent(ctx context.Context, miner address.Address,
-	// it, will start sending data via graphsync to us. This happens behind the
-	// scenes, and we dont get much introspection into this
+	// The next nonce (incrementing unique ID starting from 0) for the next voucher
+	var nonce uint64 = 0
 
-	// Now, we poll the data transfer channel status and wait until it says its accepted
-	if err := fc.waitForDealAccepted(ctx, chanid); err != nil {
-		return nil, err
-	}
+	// Used to limit prints to the console about received status
+	lastReceivedUpdate := time.Now()
 
-	var nonce uint64
-	total := abi.NewTokenAmount(0)
+	// dtRes receives either an error (failure) or nil (success) which is waited
+	// on and handled below before exiting the function
+	dtRes := make(chan error)
 
-	var voucherCount int
-	for {
-		// Now that the deal has been accepted, we sit around and wait for data
-		// to come in and for the miner to ask for more money
-		amt, nvc, done, err := fc.waitForPaymentNeeded(ctx, chanid, voucherCount)
-		if err != nil {
-			return nil, xerrors.Errorf("waitForPayment needed failed: %w", err)
-		}
+	unsubscribe := fc.dataTransfer.SubscribeToEvents(func(event datatransfer.Event, state datatransfer.ChannelState) {
+		// Lock chanid for the entire callback
+		chanidMu.Lock()
+		defer chanidMu.Unlock()
 
-		if done {
-			break
+		// Skip all events that aren't related to this channel
+		if state.ChannelID() != chanid {
+			return
 		}
 
-		voucherCount = nvc
-
-		total = types.BigAdd(total, *amt)
-
-		vres, err := fc.pchmgr.CreateVoucher(ctx, pchaddr, paych.SignedVoucher{
-			ChannelAddr: pchaddr,
-			Lane:        lane,
-			Nonce:       nonce,
-			Amount:      total,
-		})
-		if err != nil {
-			return nil, err
-		}
-
-		if types.BigCmp(vres.Shortfall, big.NewInt(0)) > 0 {
-			return nil, fmt.Errorf("not enough funds remaining in payment channel (shortfall = %s)", vres.Shortfall)
-		}
-
-		paymnt := &retrievalmarket.DealPayment{
-			ID:             proposal.ID,
-			PaymentChannel: pchaddr,
-			PaymentVoucher: vres.Voucher,
+		switch event.Code {
+		case datatransfer.NewVoucherResult:
+			switch resType := state.LastVoucherResult().(type) {
+			case *retrievalmarket.DealResponse:
+				switch resType.Status {
+				case retrievalmarket.DealStatusAccepted:
+					log.Info("deal accepted")
+
+				// Respond with a payment voucher when funds are requested
+				case retrievalmarket.DealStatusFundsNeeded:
+					log.Infof("sending payment voucher (nonce: %v, amount: %v)", nonce, resType.PaymentOwed)
+
+					totalPayment = types.BigAdd(totalPayment, resType.PaymentOwed)
+
+					vres, err := fc.pchmgr.CreateVoucher(ctx, pchAddr, paych.SignedVoucher{
+						ChannelAddr: pchAddr,
+						Lane:        pchLane,
+						Nonce:       nonce,
+						Amount:      totalPayment,
+					})
+					if err != nil {
+						dtRes <- err
+					}
+
+					if types.BigCmp(vres.Shortfall, big.NewInt(0)) > 0 {
+						dtRes <- xerrors.Errorf("not enough funds remaining in payment channel (shortfall = %s)", vres.Shortfall)
```

Comment:
on these error channel sends, there are a few things to consider. First is a thing Ive caught myself doing so many times, you replaced a return error with a channel send, and now we are no longer returning from the function (meaning that here, SendVoucher will still get called).

Second, if we get one error, then we move on in the rest of our code, what happens when the next event comes in and another error happens? It will send on the same channel, but nobody will be receiving on the other end, and this will block. I think that these callbacks happen in their own goroutine anyways, so it wont likely cause a deadlock, but it is likely to cause a resource leak, with a bunch of goroutines hanging on error sends for a long time. To get around this, we can use a select statement to wait for multiple conditions, like so:
```go
select {
case dtRes <- err:
  return
case <-doneChannel:
  return
}
```

That way, if we are 'done' we dont block forever. To implement that done channel, we can use `ctx.Done()` on our functions context, and do some cancellation on it at the top of the function:

(either before or after the tracer injection):
```go
ctx, cancel := context.WithCancel(ctx)
defer cancel()
```

Reply:
To make that cleaner, we could turn it into a helper function to wrap around the select statement:
```go
sendErr := func(e error) {
  select {
  case dtRes <- e:
  case <-ctx.Done():
  }
}
```

And then just use it like:
```
if err != nil {
  sendErr(err)
  return
}
```

Reply:
how about something like this?

```
finish := func(err error) {
  select {
  case dtRes <- err:
  default:
  }
}
```

and then, just

```
if err != nil {
  finish(err)
  return
}
```

that way it would try to push to the channel, or if it couldn't, just do nothing

Reply:
Yeah, That works too, but you need to make sure you buffer the channel so you never accidentally hit the default.
```go
dtRes := make(chan error)
```

Reply:
ok, will do. how does this change the behavior, if i may ask? i tried doing some research just now but i don't think i get what you're talking about here.

Reply:
Ah, so this is a fun one.

Channels are unbuffered by default, which means that sends will block until the other side is receiving synchronously. In the unbuffered mode, if the other side is not receiving at the exact moment of the send, a default case in a select statement would be taken. even if the receive happens a millisecond later, the case evaluation happens as we enter the select.

Now, with a buffered channel, values sent can just go into the buffer (up to the number specified) and can be received whenever. A send into a buffered channel with space *always* happens immediately, ensuring we never race to take the default path.